cms/server/ContestWebServer.py

```diff
@@ -1061,10 +1064,12 @@ def which_language(user_filename):
                              if it is not a recognized language.
 
             """
-            for source_ext, language in SOURCE_EXT_TO_LANGUAGE_MAP.iteritems():
-                if user_filename.endswith(source_ext):
-                    return language
-            return None
+
+            languageSubmitted = self.get_argument("submission_lang")
```

Comment:
According to docs, it throws 400 exception if the argument doesn't exist. This gets caught by the base handler's write_error which spits out a nicely formatted error message. I've added Tornado's explanation of the error to error.html (through the render params).

I've tested it once with Safari for mac and it seemed to work